opensafely/codelists.py

```diff
@@ -30,27 +54,162 @@ def main():
+            response.raise_for_status()
+        except Exception as e:
+            exit_with_error(
+                f"Error downloading codelist: {e}\n\n"
+                f"Check that you can access the codelist at:\n{codelist.url}"
+            )
+        codelist.filename.write_bytes(response.content)
+        new_files.add(codelist.filename)
+        key = str(codelist.filename.relative_to(codelists_dir))
+        manifest["files"][key] = {
+            "url": codelist.url,
+            "downloaded_at": f"{datetime.datetime.utcnow()}Z",
+            "sha": hash_bytes(response.content),
+        }
+    manifest_file = codelists_dir / MANIFEST_FILE
+    preserve_download_dates(manifest, manifest_file)
+    manifest_file.write_text(json.dumps(manifest, indent=2))
+    for file in old_files - new_files:
+        print(f"Deleting {file.name}")
+        file.unlink()
+    return True
+
+
+def check():
+    codelists_dir = Path.cwd() / CODELISTS_DIR
+    codelists = parse_codelist_file(codelists_dir)
+    manifest_file = codelists_dir / MANIFEST_FILE
+    if not manifest_file.exists():
+        # This is here so that switching to use this test in Github Actions
+        # doesn't cause existing repos which previously passed to start
+        # failing. It works by creating a temporary manifest file and then
+        # checking against that. Functionally, this is the same as the old test
+        # which would check against the OpenCodelists website every time.
+        if os.environ.get("GITHUB_WORKFLOW"):
+            print(
+                "==> WARNING\n"
+                "    Using temporary workaround for Github Actions tests.\n"
+                "    You should run: opensafely codelists update\n"
+            )
+            manifest = make_temporary_manifest(codelists_dir)
+        else:
+            exit_with_prompt(f"No file found at '{CODELISTS_DIR}/{MANIFEST_FILE}'.")
+    else:
+        manifest = json.loads(manifest_file.read_text())
+    all_urls = {codelist.url for codelist in codelists}
+    urls_in_manifest = {f["url"] for f in manifest["files"].values()}
+    if all_urls != urls_in_manifest:
+        exit_with_prompt(
+            f"It looks like '{CODELISTS_FILE}' has been edited but "
+            f"'update' hasn't been run."
+        )
+    all_csvs = set(f.name for f in codelists_dir.glob("*.csv"))
+    csvs_in_manifest = set(manifest["files"].keys())
+    if all_csvs != csvs_in_manifest:
+        exit_with_prompt(
+            f"It looks like CSV files have been added or deleted in the "
+            f"'{CODELISTS_DIR}' folder."
+        )
+    for filename, details in manifest["files"].items():
+        csv_file = codelists_dir / filename
+        sha = hash_bytes(csv_file.read_bytes())
+        if sha != details["sha"]:
+            exit_with_prompt(
+                f"A CSV file seems to have been modified since it was downloaded:\n"
+                f"  {CODELISTS_DIR}/{filename}\n"
+            )
```

Comment:
In each of these error cases, can you give more details (eg which files have been deleted?) and provide exact instructions for the command to run?

Reply:
The command to run is the same in every case (`opensafely codelists update`) so I made the `exit_with_prompt` function to add this to the bottom of every error message.

As to giving details of the specific files involved, I just got slightly lazy as it's a two-way set diff, two times and ultimately the corrective action is the same anyway. But you're right, it would be helpful to see what exactly has been changed.